### DISCUSSION TITLE
The http post upload binary file received incomplete data

### DISCUSSION THREAD

**Discussion by jingqunhu** [nanoframework/Home#1448] [Q&A]
It starts at about 15,000 bytes, and the rest of the data is lost
![1](https://github.com/nanoframework/Home/assets/42334968/86094ba6-fb17-4471-b537-4105c2a0555d)

**Comment by Ellerbach** [comment]
Can you please provide more context? If you are using a device with PSRAM or not, size of the memory and elements like this to help understanding the code issue. My bet: you're using a device with too limited memory and there is not enough to allocate.

**Reply by jingqunhu** [reply]
Connected to:
ESP32 (ESP32-D0WD-V3 (revision v3.1))
Features WiFi, BT, Dual Core, 240MHz, VRef calibration in efuse, Coding Scheme None
Flash size 16MB GD25Q128 from GIGADEVICE (manufacturer 0x200 device 0x16408)
PSRAM: undetermined
Crystal 40MHz
MAC B0:B2:1C:31:C5:A4

**Reply by jingqunhu** [reply]
My board is 8MB PSRAM

**Comment by jingqunhu** [comment]
/// <summary>
/// update
/// </summary>
/// <param name="e"></param>
[Route("hub/update")]
[CaseSensitive]
[Method("POST")]
public void Update(WebServerEventArgs e)
{
    try
    {
        // Get the param from the body
        Debug.WriteLine(e.Context.Request.ContentLength64.ToString());
        byte[] buff = new byte[e.Context.Request.ContentLength64];
        e.Context.Request.InputStream.Read(buff, 0, buff.Length);

        int offset = 0;
        for (int i = 0; i < buff.Length; i++)
        {
            if (buff[i].ToString() == "78" && buff[i + 1].ToString() == "70" && buff[i + 2].ToString() == "77" && buff[i + 3].ToString() == "82" && buff[i + 4].ToString() == "75")
            {
                offset = i;
                break;
            }
        }

        Debug.WriteLine(offset.ToString());

        byte[] fileBuff = new byte[buff.Length - (offset + 1)];
        Array.Copy(buff, offset, fileBuff,0,fileBuff.Length);

        for (int i = 0; i < fileBuff.Length; i++)
        {
            Debug.Write(fileBuff[i].ToString() + " ");
        }


 
        retModel.Clear();
        retModel.Add("sucess", true);
        retModel.Add("data", "");
        retModel.Add("msg", "");
        WebServer.OutPutStream(e.Context.Response, JsonSerializer.SerializeObject(retModel));

        //nanoFramework.Runtime.Native.Power.RebootDevice();
    }
    catch (Exception ex)
    {
        retModel.Clear();
        retModel.Add("sucess", true);
        retModel.Add("data", "");
        retModel.Add("msg", "");
        WebServer.OutPutStream(e.Context.Response, JsonSerializer.SerializeObject(retModel));
    }
}

**Comment by jingqunhu** [comment]
![image](https://github.com/nanoframework/Home/assets/42334968/02212822-dd15-47c9-b9d5-36d5cdc664d4)
This is a screenshot of the received binary s data compared to the original file,The left side is the webserver received, the right side is the original file binary
![image](https://github.com/nanoframework/Home/assets/42334968/b1417110-47f1-4a04-b961-6a56064b86fd)
This is the data received by the board's webserver
![image](https://github.com/nanoframework/Home/assets/42334968/900a5229-eb64-4167-bebf-b1a6c4695da2)
This is the reading of real data from the binary file through the.NET console

**Reply by jingqunhu** [reply]
@Ellerbach

**Comment by jingqunhu** [comment]
I was just trying to upload a 60kb binary file via the webserver post on the board, and this problem occurred

**Reply by josesimoes** [reply]
This is probably an issue with memory, but from native end. The memory available for the IDF heap is probably too short to hold the 60k content. 

Suggest that you do some testing with smaller files. Just to make sure the code is working with large files (some kB) but failing for files larger than... XXkB.

**Reply by Ellerbach** [reply]
And then as the memory may indeed be limited, you can try to use chunks of data rather than the full file, it's a good practice in embedded developments.

**Reply by jingqunhu** [reply]
OK, thank you.
I originally wanted to use webserver to build an html page and then upload the binary file function. Now I'm using a desktop program that uses TCP subcontracting for transmission, instead of using html pages for one-time submission. Solve the problem like this for the time being.

**Reply by Ellerbach** [reply]
Constraints devices can't be really used like non constraints ones for everything. There are indeed some limits and you may have to use different ways. I'm glad you figured out a way to go!
